### PR TITLE
#145 : Added support for COMMAND COUNT command 

### PR DESCRIPTION
**Pull request by VipinRaiP**
### Summary
Added support for  count command. Please review.
COMMAND COUNT is used to get total number of commands in dicedb.
redis documentation : https://redis.io/docs/latest/commands/command-count
This is my first PR and just getting started with the project. I have referred @yashs360's PR #150 for making the code changes. Thank you @yashs360 

### Changes

- Reused used the changes made by @yashs360  in PR #150 for adding constants for command names. Thanks @yashs360.
- Updated the eval methods as per PR #158  to support COMMAND|subcommand, thanks @kakdeykaushik
-  Added method to get the total number of commands.
- Added test and benchmark.

Benchmark results

image

Issue : #145

### CODE REVIEW

File: tests/command_count_test.go
Comment (arpitbbhayani) [inline tests/command_count_test.go:13]: Why call this subscriber? This is just a simple local TCP connection, so name it `connection` at best.

File: tests/command_count_test.go
Comment (arpitbbhayani) [inline tests/command_count_test.go:21]: instead of typecasting this to `int64` you can set commandCount as `int64` while declaring. Also, is this a good idea to test this? So, every time you add a new command, this test is bound to fail. Can you confirm how is Redis codebase testing this?

File: tests/command_count_test.go
Comment (arpitbbhayani) [inline tests/command_count_test.go:25]: Giving such generic function name is a really bad idea as this function is available throughout the package. Please reer to other benchmark functions and go's best practices and structure this better.

Also, update the output of number of allocs and time taken during the benchmar.

File: tests/command_count_test.go
Comment (arpitbbhayani) [inline tests/command_count_test.go:41]: What is the purpose of passing `200` or `2000` here? Should it just keep firing the `COMMAND COUNT` command, essentially not needing the inner for loop?

Also, the name of the command is `COMMAND COUNT` and not `COUNT`. Please go through the Redis documentation linkedin the issue description.

**Review by arpitbbhayani — CHANGES_REQUESTED** [review]

**Comment by arpitbbhayani** [comment]
@VipinRaiP in case you missed it. Let me know by where are you planning to wrap it up?

**Comment by VipinRaiP** [comment]
> @VipinRaiP in case you missed it. Let me know by where are you planning to wrap it up?

Hi @arpitbbhayani 
Sorry for the delay, i will fix these issues and push updated code over this weekend .
Please excuse

File: tests/command_count_test.go
Comment (VipinRaiP) [reply tests/command_count_test.go:13]: Sure

File: tests/command_count_test.go
Comment (VipinRaiP) [reply tests/command_count_test.go:21]: May i know where can I find the unit tests in redis codebase?

File: tests/command_count_test.go
Comment (VipinRaiP) [reply tests/command_count_test.go]: Updated the method name; Followed the benchmark standards of go.
Alloc and time taken will be printed by default when the test runs.

File: tests/command_count_test.go
Comment (VipinRaiP) [reply tests/command_count_test.go]: Updated the command name to COMMAND COUNT;

**Comment by VipinRaiP** [comment]
> @VipinRaiP in case you missed it. Let me know by where are you planning to wrap it up?

I have fixed the issues and pushed a new commit, thanks

File: tests/command_count_test.go
Comment (JyotinderSingh) [inline tests/command_count_test.go:22]: Is it a good idea to include connection establishment as a part of the benchmark code? Maybe the connection should be created once in the caller and be reused across function calls for the benchmark.

File: tests/command_count_test.go
Comment (JyotinderSingh) [inline tests/command_count_test.go:21]: This function can probably be reused for the TestCommandCount method as well.

File: tests/command_count_test.go
Comment (VipinRaiP) [reply tests/command_count_test.go]: I have updated to reuse connection.

File: tests/command_count_test.go
Comment (VipinRaiP) [reply tests/command_count_test.go]: Removed duplicate code and pushed new changes.

**Review by JyotinderSingh — APPROVED** [review]

File: core/eval.go
Comment (JyotinderSingh) [inline core/eval.go:74]: Let's instead create a fixed-size array instead of a slice since it will never grow at runtime.

File: core/eval.go
Comment (JyotinderSingh) [inline core/eval.go:935]: instead of calculating the length of the commands for every request, we can simply store it in a constant and return that. We can add a runtime check to verify that the constant's value does not get outdated.
```go
const (
    diceCommandsCount = len(diceCommands)
)

func init() {
    if len(diceCommands) != diceCommandsCount {
        panic("diceCommands slice has been modified")
    }
}
```

File: core/eval.go
Comment (JyotinderSingh) [inline core/eval.go:922]: nit: we could use errors.New() instead of fmt.Errorf().

There's not much difference, but since there are no custom args for this string we probably don't need fmt.Errorf

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]

**Comment by JyotinderSingh** [comment]
Left some minor comments.

File: core/eval.go
Comment (VipinRaiP) [reply core/eval.go:74]: sure

**Comment by gauravsarma1992** [comment]
This PR https://github.com/DiceDB/dice/pull/170 will have some conflicts with the current PR.

**Comment by VipinRaiP** [comment]
> This PR #170 will have some conflicts with the current PR.

Should this PR wait until #170  is merged and make changes as per new code?

**Comment by JyotinderSingh** [comment]
> > This PR #170 will have some conflicts with the current PR.
> 
> Should this PR wait until #170 is merged and make changes as per new code?

Hi @VipinRaiP, thanks for all the changes so far. #170 has been merged now. Please rebase your changes on top of the latest master and fix any conflicts that may be there. Once you have addressed the reviews, please resolve those comments and I will review this again.

**Comment by VipinRaiP** [comment]
Sure

**Comment by VipinRaiP** [comment]
> > > This PR #170 will have some conflicts with the current PR.
> > 
> > 
> > Should this PR wait until #170 is merged and make changes as per new code?
> 
> Hi @VipinRaiP, thanks for all the changes so far. #170 has been merged now. Please rebase your changes on top of the latest master and fix any conflicts that may be there. Once you have addressed the reviews, please resolve those comments and I will review this again.

Rebased and pushed a new commit, please review

**Review by JyotinderSingh — APPROVED** [review]

**Comment by JyotinderSingh** [comment]
Approved, will merge soon.
Could you please update the benchmark screenshot (since we have made some changes now) @VipinRaiP

**Comment by JyotinderSingh** [comment]
@VipinRaiP Looks like a unit test is failing with these changes, could you please look into it?

**Review by JyotinderSingh — APPROVED** [review]

**Pull request merged**
